xclim/locales/__init__.py

```diff
@@ -0,0 +1,394 @@
+    if locale in [av.split("-")[0] for av in available]:
+        return [av for av in available if av.split("-")[0] == locale][0]
+    return None
+
+
+def get_local_dict(locale: Union[str, Sequence[str], Tuple[str, dict]]):
+    """Return all translated metadata for a given locale.
+
+    Parameters
+    ----------
+    locale : str or sequence of str
+        IETF language tag or a tuple of the language tag and a translation dict, or
+        a tuple of the language tag and a path to a json file defining translation
+        of attributes.
+
+    Raises
+    ------
+    UnavailableLocaleError
+        If the given locale is not available.
+
+    Returns
+    -------
+    str
+        The best fitting locale string
+    dict
+        The available translations in this locale.
+    """
+    if isinstance(locale, str):
+        locale = get_best_locale(locale)
+        if locale is None:
+            raise UnavailableLocaleError(locale)
+
+        return (
+            locale,
+            json.load(pkg_resources.resource_stream(__package__, f"{locale}.json")),
+        )
+    if isinstance(locale[1], dict):
+        return locale
+    with open(locale[1]) as locf:
+        return locale[0], json.load(locf)
+
+
+class TranslatableStr(UserString):
+    """A string that can to translate certain arguments passed to format()`.
+
+    See the doc of format() for more details.
+    """
+
+    def __init__(
+        self,
+        string: str,
+        translations: Mapping[str, Sequence[str]],
+        modifiers: Sequence[str],
+    ):
+        """Initialize the translatable string.
+
+        Parameters
+        ----------
+        string : str
+            Data to potentatially translate.
+        translations : Mapping[str, Sequence[str]]
+            A mapping from a string to translate to its possible translations.
+        modifiers : Sequence[str]
+            The list of modifiers, must be the as long as the longest value of `translations`.
+        """
+        super().__init__(string)
+        self.translations = translations
+        self.modifiers = modifiers
+
+    def format(self, *args, **kwargs):
+        """Format the string by translating translatable arguments.
+
+        Does the same as str.format(), but if any of the input
+        keyword arguments is a translatable string (identified by braces { }),
+        all possible translations are passed to the formatting. The keyword is
+        removed from the arguments and replaces by all versions of `keyword` + `modifier`.
+
+        Example
+        -------
+        Let's say the string "The dog is {adj1}, the goose is {adj2}" is to be translated
+        to french and that we know that possible values of `adj` are `nice` and `evil`.
+        In french, the genre of the noun changes the adjective (cat = chat is masculine,
+        and goose = oie is feminine) so we initialize the string as:
+
+        >>> s = TranslatableStr("Le chien est {adj1_m}, l'oie est {adj2_f}",
```

Comment:
You could possibly use custom formatters instead: https://tobywf.com/2015/12/custom-formatters/

Reply:
This indeed looks better. And maybe will make it less restrictive. Thanks!